components/brave_component_updater/browser/https_upgrade_exceptions_service.cc

```diff
@@ -0,0 +1,72 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_component_updater/browser/https_upgrade_exceptions_service.h"
+
+#include <fstream>
+#include <memory>
+#include <set>
+#include <string>
+
+#include "base/files/file_path.h"
+#include "brave/components/brave_component_updater/browser/local_data_files_observer.h"
+
+#define HTTPS_UPGRADE_EXCEPTIONS_DAT_FILE "https-upgrade-exceptions-list.dat"
+#define HTTPS_UPGRADE_EXCEPTIONS_DAT_FILE_VERSION "1"
+
+namespace brave_component_updater {
+
+HttpsUpgradeExceptionsService::HttpsUpgradeExceptionsService(
+    LocalDataFilesService* local_data_files_service)
+    : LocalDataFilesObserver(local_data_files_service), is_ready_(false) {}
+
+void HttpsUpgradeExceptionsService::LoadHTTPSUpgradeExceptions(
+    const base::FilePath& install_dir) {
+  base::FilePath path =
+      install_dir.AppendASCII(HTTPS_UPGRADE_EXCEPTIONS_DAT_FILE_VERSION)
+          .AppendASCII(HTTPS_UPGRADE_EXCEPTIONS_DAT_FILE);
+  std::ifstream file_stream(path.AsUTF8Unsafe());
+  if (!file_stream.is_open()) {
+    return;
+  }
+  exceptional_domains_.clear();
+  for (std::string line; std::getline(file_stream, line);) {
+    if (!line.empty()) {
+      exceptional_domains_.insert(line);
+    }
+  }
+  is_ready_ = true;
+  return;
+}
+
+bool HttpsUpgradeExceptionsService::CanUpgradeToHTTPS(const GURL& url) {
+  if (!is_ready_) {
+    return false;
```

Comment:
I wonder if this should default to `true` if the component is not loaded yet. It won't be ready instantly on each browser startup.

Reply:
If we don't have the exceptions list loaded, then unfortunately the user could find themselves on a nonfunctional website (a website where the HTTPS version is broken but the HTTP version works). While that's not a huge number of sites, I tend to think it's better to avoid such breakage.

Reply:
@arthuredelstein could you add that as a comment? Just so future folks understand the intentional tradeoff being made?

Reply:
Sure, added